### PR TITLE
Collect network pkt timings

### PR DESCRIPTION
**Pull request by jukkar**
Collect statistics about how long the network packet is in transit from when it is allocated to when it is sent successfully by the network device driver. Print such information in network shell if CONFIG_NET_CONTEXT_TIMESTAMP is enabled.

### CODE REVIEW

**Comment by jukkar** [comment]
Changed the code to use SO_TIMESTAMPING socket option which linux is using in TX side. The SO_TIMESTAMP in linux is done for RX side only.

File: subsys/net/ip/net_stats.h [hunk 1]
```diff
@@ -308,6 +308,19 @@ static inline void net_stats_update_ipv6_mld_drop(struct net_if *iface)
 #define net_stats_update_ipv6_mld_drop(iface)
 #endif /* CONFIG_NET_STATISTICS_MLD */
 
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP) && defined(CONFIG_NET_STATISTICS)
+static inline void net_stats_update_tx_time(struct net_if *iface,
+					    u32_t start_time,
+					    u32_t end_time)
+{
+	UPDATE_STAT(iface, stats.tx_time.time_sum +=
+		    SYS_CLOCK_HW_CYCLES_TO_NS64(end_time - start_time) / 1000);
```
Comment (jukkar) [inline subsys/net/ip/net_stats.h]: Note to self: I need to fix this to check wrap-arounds.

File: subsys/net/ip/net_stats.h [hunk 2]
```diff
@@ -326,6 +339,21 @@ static inline void net_stats_update_tc_sent_priority(struct net_if *iface,
 	UPDATE_STAT(iface, stats.tc.sent[tc].priority = priority);
 }
 
+
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP)
```
Comment (jukkar) [inline subsys/net/ip/net_stats.h]: Statistics check is missing from this ifdef

File: subsys/net/ip/net_stats.h [hunk 3]
```diff
@@ -326,6 +339,21 @@ static inline void net_stats_update_tc_sent_priority(struct net_if *iface,
...
+static inline void net_stats_update_tc_tx_time(struct net_if *iface,
+					       u8_t tc,
+					       u32_t start_time,
+					       u32_t end_time)
+{
+	UPDATE_STAT(iface, stats.tc.sent[tc].time_sum +=
+		    SYS_CLOCK_HW_CYCLES_TO_NS64(end_time - start_time) / 1000);
```
Comment (jukkar) [inline subsys/net/ip/net_stats.h]: wraparound check

File: subsys/net/ip/net_stats.h [hunk 4]
```diff
@@ -349,6 +377,16 @@ static inline void net_stats_update_tc_recv_priority(struct net_if *iface,
 #define net_stats_update_tc_recv_pkt(iface, tc)
 #define net_stats_update_tc_recv_bytes(iface, tc, bytes)
 #define net_stats_update_tc_recv_priority(iface, tc, priority)
+
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP)
```
Comment (jukkar) [inline subsys/net/ip/net_stats.h]: statistics check needs to be added

**Comment by jukkar** [comment]
Rebased against latest upstream.

**Comment by jukkar** [comment]
Fixed the compilation errors.

File: subsys/net/ip/Kconfig [hunk 5]
```diff
@@ -373,7 +373,8 @@ config NET_CONTEXT_TIMESTAMP
 	bool "Add timepstamp support to net_context"
 	select NET_PKT_TIMESTAMP
 	help
-	  It is possible to timestamp outgoing packets.
+	  It is possible to timestamp outgoing packets and get information
```
Comment (pfalcon) [inline subsys/net/ip/Kconfig]: I find such descriptions confusing. It's possible to timestamp outgoing packets, good. Then, suddenly, it's possible to get information about net_pkt timings. How "timestamp outgoing packets" and "net_pkt timings" relate? Any reason this doesn't say "It is possible to timestamp outgoing packets and get information about these timestamps" ?

File: subsys/net/ip/Kconfig [hunk 6]
```diff
@@ -536,10 +537,20 @@ config NET_PKT_TIMESTAMP
 	  example in gPTP which needs to know how long it takes to send
 	  a network packet.
 
+config NET_PKT_TIMESTAMP_THREAD
```
Comment (pfalcon) [inline subsys/net/ip/Kconfig]: If, per the description, this thread is needed only for gPTP, while option name doesn't include reference to it?

File: subsys/net/ip/net_context.c [hunk 7]
```diff
@@ -1347,7 +1364,7 @@ static int context_sendto(struct net_context *context,
 		get_context_timepstamp(context, &timestamp, NULL);
 		if (timestamp) {
 			struct net_ptp_time tp = {
-				.second = k_cycle_get_32(),
+				.nanosecond = k_cycle_get_32(),
```
Comment (pfalcon) [inline subsys/net/ip/net_context.c]: Comment saying while CPU cycle counter is treated as "nanosecond"?

File: subsys/net/ip/net_stats.h [hunk 8]
```diff
@@ -308,6 +310,28 @@ static inline void net_stats_update_ipv6_mld_drop(struct net_if *iface)
 #define net_stats_update_ipv6_mld_drop(iface)
 #endif /* CONFIG_NET_STATISTICS_MLD */
 
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP) && defined(CONFIG_NET_STATISTICS)
+static inline void net_stats_update_tx_time(struct net_if *iface,
+					    u32_t start_time,
+					    u32_t end_time)
+{
+	u32_t diff = (u32_t)abs((s32_t)end_time - (s32_t)start_time);
```
Comment (pfalcon) [inline subsys/net/ip/net_stats.h]: Some MISRA pundit might say you're calling for undefined behavior here. Instead, you should subtract unsigned, and convert the result to s32_t.

File: subsys/net/ip/net_stats.h [hunk 9]
```diff
@@ -326,6 +350,33 @@ static inline void net_stats_update_tc_sent_priority(struct net_if *iface,
 	UPDATE_STAT(iface, stats.tc.sent[tc].priority = priority);
 }
 
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP) && defined(CONFIG_NET_STATISTICS)
+static inline void net_stats_update_tc_tx_time(struct net_if *iface,
+					       u8_t tc,
+					       u32_t start_time,
+					       u32_t end_time)
+{
+	u32_t diff = (u32_t)abs((s32_t)end_time - (s32_t)start_time);
```
Comment (pfalcon) [inline subsys/net/ip/net_stats.h]: Potential undefined behavior on signed int overflow.

**Review by pfalcon — CHANGES_REQUESTED** [review]
Watched one of the commits, have a number of questions. Still might be false positives, but let me for give -1, so these questions were considered and didn't slip.

File: subsys/net/ip/Kconfig (hunk 6 above)
Comment (jukkar) [reply subsys/net/ip/Kconfig]: gPTP needs this support, but timestamping does not need gptp. Currently only gptp uses or needs this one but I do not see a reason to mention gptp in the option name.

**Comment by jukkar** [comment]
Updated according to comments.

**Comment by jukkar** [comment]
ping @pfalcon, please re-review

File: subsys/net/ip/net_context.c [hunk 10]
```diff
@@ -1347,7 +1364,24 @@ static int context_sendto(struct net_context *context,
...
+				 * net_if.c:net_if_tx()
+				 *
+				 * The net_pkt timestamp field is used in two
+				 * roles here:
+				 * 1) To calculate how long it takes the packet
+				 *    from net_context to be sent by the
+				 *    network device driver.
+				 * 2) gPTP enabled Ethernet device driver will
+				 *    use the value to tell gPTP what time the
+				 *    packet was send.
```
Comment (pfalcon) [inline subsys/net/ip/net_context.c]: Minor typo:

>  was send

Should be "was sent".

File: subsys/net/ip/net_if.c [hunk 11]
```diff
@@ -141,6 +141,13 @@ static bool net_if_tx(struct net_if *iface, struct net_pkt *pkt)
 	struct net_context *context;
 	int status;
 
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP)
+	/** Timestamp of the current network packet sent */
+	struct net_ptp_time start_timestamp;
+	u32_t curr_time = 0;
+	u8_t pkt_priority;
```
Comment (pfalcon) [inline subsys/net/ip/net_if.c]: I'm surprised to see "pkt_priority" to be wrapped in "CONFIG_NET_CONTEXT_TIMESTAMP".

File: subsys/net/ip/net_if.c [hunk 12]
```diff
@@ -157,7 +164,27 @@ static bool net_if_tx(struct net_if *iface, struct net_pkt *pkt)
 			net_pkt_set_queued(pkt, false);
 		}
 
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP)
+		if (context) {
+			if (net_context_get_timestamp(context, pkt,
+						      &start_timestamp) < 0) {
+				start_timestamp.nanosecond = 0;
+			} else {
+				pkt_priority = net_pkt_priority(pkt);
```
Comment (pfalcon) [inline subsys/net/ip/net_if.c]: And what happens on the other branch? pkt_priority stays uninitialized?

File: subsys/net/ip/net_if.c [hunk 13]
```diff
@@ -175,6 +202,20 @@ static bool net_if_tx(struct net_if *iface, struct net_pkt *pkt)
 			context, status);
 
 		net_context_send_cb(context, status);
+
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP)
+		if (status >= 0 && start_timestamp.nanosecond &&
+		    curr_time > 0) {
+			/* So we know now how long the network packet was in
+			 * transit from where it was allocated to when we
```
Comment (pfalcon) [inline subsys/net/ip/net_if.c]: Minor: more logically would be to write "from when ... to when ..." (here "where" was used in the first case).

File: subsys/net/ip/net_stats.h [hunk 14]
```diff
@@ -308,6 +310,28 @@ static inline void net_stats_update_ipv6_mld_drop(struct net_if *iface)
 #define net_stats_update_ipv6_mld_drop(iface)
 #endif /* CONFIG_NET_STATISTICS_MLD */
 
+#if defined(CONFIG_NET_CONTEXT_TIMESTAMP) && defined(CONFIG_NET_STATISTICS)
+static inline void net_stats_update_tx_time(struct net_if *iface,
+					    u32_t start_time,
+					    u32_t end_time)
+{
+	u32_t diff = (u32_t)abs((s32_t)(end_time - start_time));
```
Comment (pfalcon) [inline subsys/net/ip/net_stats.h]: Yeah, so this doesn't look correct after all. If we know that "end_time" is really end time, i.e. happened after start_time, then the formula to calculate duration between them, taking into account possible wraparound is merely `end_time - start_time`.

**Review by pfalcon — DISMISSED** [review]
@jukkar, Thanks for updates, but I still can't get beyond the very 1st commit with my review.

File: subsys/net/ip/net_if.c (hunk 12 above)
Comment (jukkar) [reply subsys/net/ip/net_if.c]: That branch is not entered in that case as start_timestamp.nanosecond == 0

**Comment by jukkar** [comment]
Updated according to comments.

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**